### PR TITLE
publish api package to github registry

### LINKED ISSUE #2089: Publish API packages to Github npm registry
### Context

Part of #2087 
:warning: Blocked by #2088 

### Task

- Check we have the appropriate configuration for packaging and distribution (esp. package.json and tsconfig.json); refer to the appropriate documentation.
- Follow the Github documentation to publish the packages to the [npm registry](https://docs.github.com/en/packages/working-with-a-github-packages-registry/working-with-the-npm-registry)
  - :warning: Make sure to publish to Github packages rather than npmjs.com!
  - :warning: :rotating_light: Make sure to change the `version` in `package.json` to some unique value, e.g. `0.0.0-df234014`; any given version of a package can only be published _once_!.

### Acceptance criteria

_On a temporary branch_, change the ts-tests to point to the published packages. Check that `yarn install` works and the tests run successfully. **Do not merge this change to dev** before we've set up the rest of the pipeline (otherwise, changes to the Rust code will break the tests and would require manual publishing to fix).

### LINKED ISSUE #2090: Automate API package publication
### Context

Part of #2087 
:warning: Blocked by #2089 

### Task

- Create a script (Makefile step? Yarn script? Shell script? TBD) that takes the currently built version of the API packages and publishes it to the npm registry. Refer to the outcome of #2089 for procedure details.
  - NOTE: the script should take a branch name as parameter, and use that to **tag** the published versions.
  - NOTE: publishing will fail if the version (as defined in `package.json`) already exists; do we need to handle this somehow?

### Acceptance criteria

Use the script to publish a new version of the packages.

### LINKED ISSUE #2091: Github action for API package publication
### Context

Part of #2087 
:warning: Blocked by #2090 

### Task

- Create a Github action to trigger publication of the API packages with their given version.
  - If the given version was already published, check the checksum of the published version:
    - If it matches the current one, continue normally
    - Otherwise, fail with an error (the version number needs to be bumped)
  - Tag the given version with the current branch name
- NOTE: We should _not_ bump the version number automatically!
- The action will need to authenticate to Github using a token. We should set up a shared "CI/system" user for this (i.e. **not** an individual developer's token).
  - The token needs to be added to the repo secrets
  - the token needs to be written into the `~/.npmrc` file when running the action

### Acceptance criteria

A new version of the packages successfully published and tagged _by another team member_ via the CI action

### PR DESCRIPTION
**Pull request by 0xverin**
resolves: #2089 
resolves: #2090 
resolves: #2091

### CODE REVIEW

**Comment by 0xverin** [comment]
~~Not ready for review yet, still needs some testing.~~

File: .github/workflows/release-client-api.yml
Comment (arieljbirnbaum) [inline .github/workflows/release-client-api.yml]: no `--tag dev` ?

File: .github/workflows/release-client-api.yml
Comment (arieljbirnbaum) [inline .github/workflows/release-client-api.yml]: How do we ensure here that the worker is running (with the expected ports)?
For reference, the ts-tests use docker-compose to set, e.g.: `depends_on.litentry-worker-1.condition: service_healthy`; see `tee-worker/docker/lit-di-vc-test.yml`

File: .github/workflows/release-client-api.yml
Comment (0xverin) [reply .github/workflows/release-client-api.yml]: So, do we need three tags? `dev`, `staging`, and `latest`, right?

File: .github/workflows/release-client-api.yml
Comment (0xverin) [reply .github/workflows/release-client-api.yml]: Emmm, I'm directly using metadata to build here (assuming that the metadata on the branch is correct). If follow your suggestion, do we need to update the metadata and then build?

File: .github/workflows/release-client-api.yml
Comment (arieljbirnbaum) [reply .github/workflows/release-client-api.yml]: Not sure we need a `latest` tag; I'd say one tag per repo mainline and that's it. Note that [this](https://github.com/litentry/litentry-parachain/issues/2130) will probably also become a mainline :thinking:

File: .github/workflows/release-client-api.yml
Comment (arieljbirnbaum) [reply .github/workflows/release-client-api.yml]: Sort of. The ultimate goal is to get rid of the committed metadata and use a fresh one each time we build the package; see #2092

**Comment by 0xverin** [comment]
**Release process**:

~~1. Wait for all the CI tests to pass.~~

~~1. Manually trigger the 'release-api-package' action (choose the branch to run it on).~~

**Code flow**:

1. ~`Version check`: Verify if there's a conflict between the version in `package.json` and the version in GitHub Package. If there's a conflict, it fails; otherwise, proceed to the next step.~(seems a bit redundant; it throws an error of version conflict  when running `npm publish`.)

2. ~~`Start the worker`: Launch the parachain and worker (referencing the configuration in `lit-ii-identity-test.yml` perhaps?).~~

3. ~~`Compile the package`: Go to the 'client-api' directory, update the metadata for `sidechain` and `parachain`, and then build both.~~

4. ~~Publish the package.~~

Q: ~~For code flow's step 2, when we choose a branch to run, like `staging`, does it run the worker in the `staging` environment? cc @grumpygreenguy~~

~~Please feel free to share your thoughts. Thanks! :p~~

**Comment by Kailai-Wang** [comment]
Is it ready to review? I see both "WIP" and "ready to review" - thus confused

**Comment by 0xverin** [comment]
> Is it ready to review? I see both "WIP" and "ready to review" - thus confused

not yet. I'm taking this [test](https://github.com/users/Verin1005/packages/npm/package/verin-apple-api), and I will let you know the review when the test is done. :)

**Comment by 0xverin** [comment]
The completion of this PR addresses the tasks in issues #2090 and #2091 . For #2092 and #2093 , new PR will be needed to complete them.

I believe this PR is ready for review. It is crucial for the package release, and once it's merged, we can proceed with the package release. After that, idhub's colleagues  can use the package we've released, which is a significant step.

**Test**:
I have already tried to publish with this GHA and the publish log is [here](https://github.com/Verin1005/api-test/actions/runs/6497236671/job/17645878967)(my own personal repo)

**Note**:
We need to set up a token in advance for the release, please see [comment](https://github.com/litentry/litentry-parachain/issues/2091#issuecomment-1755440857).

**Comment by Kailai-Wang** [comment]
> We need to set up a token in advance for the release

If you have the secret, I can set it in the repo

File: .github/workflows/release-api-package.yml
Comment (Kailai-Wang) [inline .github/workflows/release-api-package.yml]: Do we consider any additional triggers? But at the moment it's fine to have manual trigger only

**Comment by 0xverin** [comment]
> > We need to set up a token in advance for the release
> 
> If you have the secret, I can set it in the repo

From the [description](https://github.com/litentry/litentry-parachain/issues/2091#issue-1875658230), it seems we might need set a company token, not an individual developer's token.

File: .github/workflows/release-api-package.yml
Comment (Kailai-Wang) [inline .github/workflows/release-api-package.yml:1]: I suggest having a more specific name, like `release-ts-api-package` or including the name of package here, as we __could__ publish rust crates (packages) too

File: .github/workflows/release-api-package.yml
Comment (Kailai-Wang) [inline .github/workflows/release-api-package.yml]: We have at least v4 now :)

File: .github/workflows/release-api-package.yml
Comment (Kailai-Wang) [inline .github/workflows/release-api-package.yml]: I suggest a more specific name, `RELEASE_TOKEN` can be too generic in the repo's scope

File: .github/workflows/release-api-package.yml
Comment (Kailai-Wang) [inline .github/workflows/release-api-package.yml]: env can be set at workflow level

File: .github/workflows/release-api-package.yml
Comment (Kailai-Wang) [inline .github/workflows/release-api-package.yml]: I suggest defining the condition at step level, maybe that is clearer, like (untested):

```
- name: Build and publish parachain and sidechain
   if:  github.ref_name == 'dev' || github.ref_name == 'staging'
   run: |
     cd tee-worker/client-api/parachain-api
     pnpm build
     npm publish --tag ${{ github.ref_name }}
     ....
```

And maybe merge the sidechain build to this step too, because they share the same if-condition.
If you really want a separation, you can use `echo "::group::...` and `echo "::endgroup::"`

**Review by Kailai-Wang — CHANGES_REQUESTED** [review]
Thanks! I have a few comments

File: .github/workflows/release-api-package.yml
Comment (arieljbirnbaum) [inline .github/workflows/release-api-package.yml]: Shouldn't we rather `corepack enable pnpm` here to make sure it uses the version set in the `package.json`? :thinking:

File: .github/workflows/release-api-package.yml
Comment (arieljbirnbaum) [inline .github/workflows/release-api-package.yml]: :nail_care: 

```suggestion
      - name: Build and Publish Parachain API
```

File: .github/workflows/release-api-package.yml
Comment (arieljbirnbaum) [inline .github/workflows/release-api-package.yml]: :nail_care: 
```suggestion
      - name: Build and Publish Sidechain API
```

File: .github/workflows/release-api-package.yml
Comment (arieljbirnbaum) [inline .github/workflows/release-api-package.yml]: Q: as long as the trigger is manual, do we really need to disallow publishing/tagging from another branch? It's not too farfetched that we'd need some side version for a special integration or whatnot :thinking:

File: .github/workflows/release-api-package.yml
Comment (0xverin) [reply .github/workflows/release-api-package.yml]: We may consider other triggers after manually releasing stability. :)

File: .github/workflows/release-api-package.yml
Comment (0xverin) [reply .github/workflows/release-api-package.yml:1]: Thanks! changed!

File: .github/workflows/release-api-package.yml
Comment (0xverin) [reply .github/workflows/release-api-package.yml]: Changed, and also changed actions/setup-node to version v3.

File: .github/workflows/release-api-package.yml
Comment (0xverin) [reply .github/workflows/release-api-package.yml]: Thanks! fixed.

File: .github/workflows/release-api-package.yml
Comment (0xverin) [reply .github/workflows/release-api-package.yml]: changed!

File: .github/workflows/release-api-package.yml
Comment (0xverin) [reply .github/workflows/release-api-package.yml]: fixed and simplified a bit.

File: .github/workflows/release-api-package.yml
Comment (0xverin) [reply .github/workflows/release-api-package.yml]: Thanks! fixed!

File: .github/workflows/release-api-package.yml
Comment (0xverin) [reply .github/workflows/release-api-package.yml]: You make a valid point, the previous addition was made considering that this is a manually entered parameter. 
Removed now. >3

File: .github/workflows/release-api-package.yml
Comment (Kailai-Wang) [reply .github/workflows/release-api-package.yml:1]: I don't see it's changed 😂

File: .github/workflows/release-api-package.yml
Comment (Kailai-Wang) [inline .github/workflows/release-api-package.yml]: typo: parach**ain**

File: .github/workflows/release-api-package.yml
Comment (Kailai-Wang) [inline .github/workflows/release-api-package.yml]: Maybe you want to say "sidechain"?

File: .github/workflows/release-api-package.yml
Comment (Kailai-Wang) [reply .github/workflows/release-api-package.yml:1]: I mean the file name and workflow name

**Review by Kailai-Wang — APPROVED** [review]
Looks mostly fine from my side, thanks!

File: .github/workflows/release-api-package.yml
Comment (0xverin) [reply .github/workflows/release-api-package.yml]: Thanks. fixed.

**Pull request merged**
